nobl9/resource_budgetadjustment.go

```diff
@@ -0,0 +1,289 @@
+			AttributePath: path,
+		})
+	}
+	return diags
+}
+
+func resourceBudgetAdjustmentApply(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(ProviderConfig)
+	client, ds := getClient(config)
+	if ds != nil {
+		return ds
+	}
+
+	budgetAdjustment := marshalBudgetAdjustment(d)
+
+	if err := retry.RetryContext(ctx, d.Timeout(schema.TimeoutCreate)-time.Minute, func() *retry.RetryError {
+		err := client.Objects().V1().Apply(ctx, []manifest.Object{budgetAdjustment})
+		if err != nil {
+			if errors.Is(err, errConcurrencyIssue) {
+				return retry.RetryableError(err)
+			}
+			return retry.NonRetryableError(err)
+		}
+		return nil
+	}); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(budgetAdjustment.Metadata.Name)
+	return resourceBudgetAdjustmentRead(ctx, d, meta)
+}
+
+func marshalBudgetAdjustment(d *schema.ResourceData) *budgetadjustment.BudgetAdjustment {
+	firstEventStart, _ := time.Parse(time.RFC3339, d.Get("first_event_start").(string))
+
+	adjustment := budgetadjustment.New(
+		budgetadjustment.Metadata{
+			Name:        d.Get("name").(string),
+			DisplayName: d.Get("display_name").(string),
+		},
+		budgetadjustment.Spec{
+			Description:     d.Get("description").(string),
+			FirstEventStart: firstEventStart,
+			Duration:        d.Get("duration").(string),
+			Rrule:           d.Get("rrule").(string),
+			Filters:         marshalFilters(d.Get("filters")),
+		})
+
+	return &adjustment
+}
+
+func marshalFilters(filters interface{}) budgetadjustment.Filters {
+	filtersSet := filters.(*schema.Set)
+	if filtersSet.Len() == 0 {
+		return budgetadjustment.Filters{}
+	}
+	slos := filtersSet.List()[0].(map[string]interface{})["slos"].(*schema.Set)
+	slosList := slos.List()[0].(map[string]interface{})["slo"].([]interface{})
+	sloRef := make([]budgetadjustment.SLORef, 0, len(slosList))
+	for _, filter := range slosList {
+		f := filter.(map[string]interface{})
+		slo := budgetadjustment.SLORef{
+			Name:    f["name"].(string),
+			Project: f["project"].(string),
+		}
+		sloRef = append(sloRef, slo)
+	}
+
+	return budgetadjustment.Filters{
+		SLOs: sloRef,
+	}
+}
+
+func resourceBudgetAdjustmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(ProviderConfig)
+	client, ds := getClient(config)
+	if ds != nil {
+		return ds
+	}
+
+	budgetAdjustments, err := client.Objects().V1().GetBudgetAdjustments(ctx, v1.GetBudgetAdjustmentRequest{
+		Names: []string{d.Id()},
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return unmarshalBudgetAdjustment(d, budgetAdjustments)
+}
+
+func unmarshalBudgetAdjustment(d *schema.ResourceData, objects []budgetadjustment.BudgetAdjustment) diag.Diagnostics {
+	if len(objects) != 1 {
+		d.SetId("")
+		return nil
+	}
+	object := objects[0]
+	var diags diag.Diagnostics
+	var err error
+
+	err = d.Set("name", object.Metadata.Name)
+	diags = appendError(diags, err)
```

Comment:
wouldn't it be simpler to simply write:

```suggestion
	diags = appendError(diags, d.Set("name", object.Metadata.Name))
```